controllers/volumereplicationgroup_controller.go

```diff
@@ -898,15 +898,98 @@ func (v *VRGInstance) createVR(vrNamespacedName types.NamespacedName, state volr
 			Namespace: vrNamespacedName.Namespace,
 		},
 		Spec: volrep.VolumeReplicationSpec{
-			DataSource: &corev1.TypedLocalObjectReference{
+			DataSource: corev1.TypedLocalObjectReference{
 				Kind: "PersistentVolumeClaim",
 				Name: vrNamespacedName.Name,
 			},
-			State: state,
+			ReplicationState: state,
+
+			VolumeReplicationClass: v.instance.Spec.VolumeReplicationClass,
 		},
 	}
 
-	return v.reconciler.Create(v.ctx, cr)
+	// Let VRG receive notification for any changes to VolumeReplication CR
+	// created by VRG.
+	if err := ctrl.SetControllerReference(v.instance, cr, v.reconciler.Scheme); err != nil {
+		return fmt.Errorf("failed to set owner reference to VolumeReplication resource (%s/%s)",
+			cr.Name, cr.Namespace)
+	}
+
+	v.log.Info("Creating VolumeReplication resource", "resource", cr)
+
+	if err := v.reconciler.Create(v.ctx, cr); err != nil {
+		return fmt.Errorf("failed to create VolumeReplication resource (%s)", vrNamespacedName)
+	}
+
+	return nil
+}
+
+func (v *VRGInstance) checkVRStatus(volRep *volrep.VolumeReplication) error {
+	// When the generation in the status is updated, VRG would get a reconcile
+	// as it owns VolumeReplication resource.
+	if volRep.Generation != volRep.Status.ObservedGeneration {
+		v.log.Info("Generation from the resource and status not same")
+
+		return nil
+	}
+
+	switch {
+	case v.instance.Spec.ReplicationState == volrep.Primary:
+		return v.validateVRStatusAsPrimary(volRep)
+	case v.instance.Spec.ReplicationState == volrep.Secondary:
+		return v.validateVRStatusAsSecondary(volRep)
+	case v.instance.Spec.ReplicationState == volrep.Resync:
+		return fmt.Errorf("resync is not supported for VRG")
+	default:
+		return fmt.Errorf("invalid Replication State %s for VolumeReplicationGroup (%s:%s)",
+			string(v.instance.Spec.ReplicationState), v.instance.Name, v.instance.Namespace)
+	}
+}
+
+// So, VolumeReplication controller reflects the status of a VolRep resource as primary
+// only after the promotion of the associated storage volume to primary is successful.
+// Hence, checking whether the primary state is reflected in the status is sufficient
+// to verify whether the VolRep resource is in a healthy state not. However, checking
+// conditions of VolRep status may be required for maintaining correct status of VRG
+// when VRG status is implemented.
+//
+// TODO: If checking conditions of VolRep is necessary for VRG status, add that logic
+//       here when VRG status is implemented.
+func (v *VRGInstance) validateVRStatusAsPrimary(volRep *volrep.VolumeReplication) error {
+	if volRep.Status.State != volrep.PrimaryState {
+		v.log.Info("State is primary and status is not primary")
+
+		return fmt.Errorf("primary VolumeReplication resource status not same %s/%s (status: %s)",
+			volRep.Name, volRep.Namespace, volRep.Status.State)
```

Comment:
Should  this be an error?  We may have just requested a VR state change to result in this case, isn't it?  My question also applies to the next method.

On the other hand, I don't know if we validate the VR spec, which may be more useful to protect from inadvertent change by the user.

Reply:
True, this needs to be co-ordinated with status update in the subsequent patch, where we update VRG status as no-ready if VR states do not match.

Currently we are just checking if desired state is the actual state if we did not make any changes to VR, I would hence say this is a filler as of now in this PR.

Reply:
So, one thing here is that this function validateVRStatusAsPrimary() is called from the function checkVRStatus. And checkVRStatus first compares the generation number for the resource and its status. So I thought that if the generation number match, then the state and status should match?

I agree that if the generation number is not same, then probably line 963 (appllies to the next function validateVRStatusAsSecondary also)  should not be an error. But, the generation number comparison is done first. If they match, then only VRG expects the state and status of VR to match.

Reply:
I believe I was not clear earlier. I was arguing against returning an error from this method, as I don't believe it is an error.  The concept of validating VR status does not make sense to me; because, all we need to do is roll it up into VRG status.  Validating VR spec makes sense, which we already do it, according to Raghu's validation: when VR state is inadvertently (manually) changed (say, to resync), VRG restores it to the required state.

Reply:
ACK. Yes, the VR status can be rolled up into VRG status. And if the status of VR changes (because, VR controller was able to successfully set it to the desired state), then VRG would receive a notification and can reconcile.

Will be removing that returning of error in validateVRStatusAsPrimary and validateVRStatusAsSecondary in a subsequent PR to handle VRG status.